tlair/src/main/scala/at/forsyte/apalache/tla/lir/TlaExLevelFinder.scala

```diff
@@ -0,0 +1,39 @@
+package at.forsyte.apalache.tla.lir
+
+import at.forsyte.apalache.tla.lir.oper.{TlaActionOper, TlaTempOper}
+
+/**
+ * This class computes the level of a TLA+ expression. See Lamport. Specifying Systems, p. 322.
+ *
+ * @param nameLevel a function that returns the level of a name.
+ * @author Igor Konnov
+ */
+class TlaExLevelFinder(nameLevel: String => TlaLevel) {
+  def apply(ex: TlaEx): TlaLevel = find(ex)
+
+  private def find: TlaEx => TlaLevel = {
+    case NameEx(name) =>
+      nameLevel(name)
+
+    case OperEx(op, args @ _*)
+        if op == TlaActionOper.prime || op == TlaActionOper.enabled
+          || op == TlaActionOper.unchanged || op == TlaActionOper.stutter
+          || op == TlaActionOper.nostutter || op == TlaActionOper.composition =>
+      TlaLevelAction.join(args.map(find))
+
+    case OperEx(op, args @ _*)
+        if op == TlaTempOper.box || op == TlaTempOper.diamond || op == TlaTempOper.leadsTo
+          || op == TlaTempOper.weakFairness || op == TlaTempOper.strongFairness
+          || op == TlaTempOper.guarantees || op == TlaTempOper.AA || op == TlaTempOper.EE =>
+      TlaLevelTemporal.join(args.map(find))
+
+    case OperEx(_, args @ _*) =>
+      TlaLevelConst.join(args.map(find))
```

Comment:
It could just be a matter of taste, but IMO this approach is a bit easier to read (assuming it's feasible):

```suggestion
case OperEx(op, args @ _*) =>
    val ctor = op match {
        case TlaActionOper.prime | TlaActionOper.enabled
            | TlaActionOper.unchanged | TlaActionOper.stutter
            | TlaActionOper.nostutter | TlaActionOper.composition => TlaLevelAction
        case TlaTempOper.box | TlaTempOper.diamond
            | TlaTempOper.leadsTo | TlaTempOper.weakFairness
            | TlaTempOper.strongFairness | TlaTempOper.guarantees
            | TlaTempOper.AA | TlaTempOper.EE => TlaLevelTemporal
        case _ => TlaLevelConst
    }
    ctor.join(args.map(find))
```

Reply:
Good point. I think I will just introduce a map.